File: 0098_Validate_Binary_Search_Tree/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
  public:
    bool isValidBST(TreeNode *root) { return dfs(root, nullptr, nullptr); }

    bool dfs(TreeNode *root, TreeNode *lower, TreeNode *upper) {
        if (root == nullptr) {
            return true;
        }
        if (lower != nullptr && lower->val >= root->val) {
            return false;
        }
        if (upper != nullptr && upper->val <= root->val) {
            return false;
        }
        return dfs(root->left, lower, root) && dfs(root->right, root, upper);
    }
};

File: 0098_Validate_Binary_Search_Tree/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// T = O(n), S = O(n)
func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return dfs(root, nil, nil)
}

func dfs(root *TreeNode, minNode *TreeNode, maxNode *TreeNode) bool {
	if root == nil {
		return true
	}
	if minNode != nil && root.Val <= minNode.Val {
		return false
	}
	if maxNode != nil && root.Val >= maxNode.Val {
		return false
	}
	return dfs(root.Left, minNode, root) && dfs(root.Right, root, maxNode)
}